File: src/engine/pinyin.ts
import type { CharData, IdiomInput, MarkState } from '../entity/idiom';

export const INITIALS = [
  'zh', 'ch', 'sh', 'b', 'p', 'm', 'f', 'd', 't', 'n', 'l', 'g', 'k', 'h', 'j', 'q', 'x', 'r',
  'z', 'c', 's', 'y', 'w',
];

const TONE_CHARS: Record<string, string> = {
  'ā': 'a1', 'á': 'a2', 'ǎ': 'a3', 'à': 'a4',
  'ē': 'e1', 'é': 'e2', 'ě': 'e3', 'è': 'e4',
  'ī': 'i1', 'í': 'i2', 'ǐ': 'i3', 'ì': 'i4',
  'ō': 'o1', 'ó': 'o2', 'ǒ': 'o3', 'ò': 'o4',
  'ū': 'u1', 'ú': 'u2', 'ǔ': 'u3', 'ù': 'u4',
  'ǖ': 'v1', 'ǘ': 'v2', 'ǚ': 'v3', 'ǜ': 'v4',
};

const TONE_MARKS: Record<string, string[]> = {
  a: ['', 'ā', 'á', 'ǎ', 'à'],
  e: ['', 'ē', 'é', 'ě', 'è'],
  i: ['', 'ī', 'í', 'ǐ', 'ì'],
  o: ['', 'ō', 'ó', 'ǒ', 'ò'],
  u: ['', 'ū', 'ú', 'ǔ', 'ù'],
  v: ['', 'ǖ', 'ǘ', 'ǚ', 'ǜ'],
};

const MARKED_VOWEL = /[āáǎàēéěèīíǐìōóǒòūúǔùǖǘǚǜ]/;

function toneVowelIndex(base: string): number {
  const s = base.toLowerCase().replace(/v/g, 'ü');
  if (s.includes('a')) return s.indexOf('a');
  if (s.includes('e')) return s.indexOf('e');
  if (s.includes('ou')) return s.indexOf('o');
  if (s.includes('iu')) return s.indexOf('u');
  if (s.includes('ui')) return s.indexOf('i');
  for (let i = s.length - 1; i >= 0; i--) {
    if ('aeiouü'.includes(s[i])) return i;
  }
  return -1;
}

export function markSyllable(syl: string): string {
  const raw = syl.trim();
  if (!raw || MARKED_VOWEL.test(raw)) return raw;
  const m = raw.toLowerCase().match(/^([a-z:]+)([1-5])$/i);
  if (!m) return raw;
  let base = m[1].replace(/v/g, 'ü');
  const tone = parseInt(m[2], 10);
  if (tone === 5) return base;
  const idx = toneVowelIndex(base);
  if (idx < 0) return base;
  const ch = base[idx];
  const key = ch === 'ü' ? 'v' : ch;
  const marks = TONE_MARKS[key];
  if (!marks) return base;
  return base.slice(0, idx) + marks[tone] + base.slice(idx + 1);
}

export function normalizePinyinInput(pinyin: string): string {
  return pinyin.trim().split(/\s+/).filter(Boolean).map(markSyllable).join(' ');
}

export function normalizeEntry(item: Partial<IdiomInput>): IdiomInput | null {
  if (!item || typeof item !== 'object') return null;
  const word = String(item.word || '').trim();
  const pinyin = normalizePinyinInput(String(item.pinyin || ''));
  if (!word || !pinyin) return null;
  return {
    word,
    pinyin,
    explanation: String(item.explanation || '').trim(),
  };
}

export function normalizeCorpus(data: unknown): IdiomInput[] {
  if (!Array.isArray(data)) throw new Error('词库必须是 JSON 数组');
  const out: IdiomInput[] = [];
  for (const item of data) {
    const e = normalizeEntry(item as Partial<IdiomInput>);
    if (e) out.push(e);
  }
  return out;
}

function stripTone(syl: string): { base: string; tone: number } {
  let s = syl.toLowerCase().replace(/ü/g, 'v');
  let tone = 0;
  const digit = s.match(/([1-5])$/);
  if (digit) {
    tone = parseInt(digit[1], 10);
    s = s.slice(0, -1);
  }
  for (const [ch, rep] of Object.entries(TONE_CHARS)) {
    if (s.includes(ch)) {
      const m = rep.match(/(.)([1-5])?/);
      s = s.replace(ch, m?.[1] || '');
      if (m?.[2]) tone = parseInt(m[2], 10);
      break;
    }
  }
  return { base: s, tone: tone || 5 };
}

function splitInitial(py: string): { sm: string; ym: string } {
  const lower = py.toLowerCase();
  const sorted = [...INITIALS].sort((a, b) => b.length - a.length);
  for (const init of sorted) {
    if (lower.startsWith(init)) {
      return { sm: init, ym: lower.slice(init.length) };
    }
  }
  return { sm: '', ym: lower };
}

export function parseSyllable(syl: string): Omit<CharData, 'c'> {
  const { base, tone } = stripTone(syl.trim());
  const { sm, ym } = splitInitial(base);
  return { py: base, sm, ym, tone };
}

export function charsFromWordPinyin(word: string, pinyin: string): CharData[] | null {
  const chars = [...word.replace(/\s/g, '')];
  const syls = normalizePinyinInput(pinyin).split(/\s+/).filter(Boolean);
  if (chars.length !== syls.length) return null;
  return chars.map((c, i) => {
    const p = parseSyllable(syls[i]);
    return { c, ...p };
  });
}

export function toneDigit(tone: number): string {
  if (tone >= 1 && tone <= 5) return String(tone);
  return '5';
}

export function charsToDigitPinyin(chars: CharData[]): string {
  return chars
    .map((ch) => {
      const tone = ch.tone >= 1 && ch.tone <= 5 ? ch.tone : '';
      return ch.py + (tone && tone !== 5 ? tone : tone === 5 ? '5' : '');
    })
    .join(' ');
}

export function formatPinyinAsDigit(pinyin: string): string {
  const syls = pinyin.split(/\s+/).filter(Boolean);
  return syls
    .map((syl) => {
      const p = parseSyllable(syl);
      if (p.tone >= 1 && p.tone <= 4) return p.py + p.tone;
      if (p.tone === 5) return p.py + '5';
      return p.py;
    })
    .join(' ');
}

export function toneLabel(tone: number): string {
  if (tone >= 1 && tone <= 4) return String(tone);
  return '轻';
}

export const DEDUCE_ATTRS = [
  { key: 'char' as const, label: '字', get: (ch: CharData) => ch.c },
  { key: 'sm' as const, label: '声母', get: (ch: CharData) => ch.sm || '零' },
  { key: 'ym' as const, label: '韵母', get: (ch: CharData) => ch.ym || '—' },
  { key: 'tone' as const, label: '声调', get: (ch: CharData) => toneLabel(ch.tone) },
];

export function charOwnsPronunciation(marks: { char?: MarkState | string | null }): boolean {
  return marks.char === 'hit' || marks.char === 'present';
}

export function syncPronunciationMarksForChar(marksAt: {
  char: MarkState | string | null;
  sm: MarkState | string | null;
  ym: MarkState | string | null;
  tone: MarkState | string | null;
}): void {
  if (marksAt.char === 'hit' || marksAt.char === 'present') {
    marksAt.sm = null;
    marksAt.ym = null;
    marksAt.tone = null;
  }
}

File: src-tauri/src/pinyin.rs
use crate::entity::idiom::{CharData, IdiomInput};
use serde::{Deserialize, Serialize};

const INITIALS: &[&str] = &[
    "zh", "ch", "sh", "b", "p", "m", "f", "d", "t", "n", "l", "g", "k", "h", "j", "q", "x", "r",
    "z", "c", "s", "y", "w",
];

#[derive(Debug, Clone, Serialize, Deserialize)]
struct ToneChar {
    base: char,
    tone: Option<u8>,
}

fn tone_chars() -> &'static [(&'static str, ToneChar)] {
    &[
        ("ā", ToneChar { base: 'a', tone: Some(1) }),
        ("á", ToneChar { base: 'a', tone: Some(2) }),
        ("ǎ", ToneChar { base: 'a', tone: Some(3) }),
        ("à", ToneChar { base: 'a', tone: Some(4) }),
        ("ē", ToneChar { base: 'e', tone: Some(1) }),
        ("é", ToneChar { base: 'e', tone: Some(2) }),
        ("ě", ToneChar { base: 'e', tone: Some(3) }),
        ("è", ToneChar { base: 'e', tone: Some(4) }),
        ("ī", ToneChar { base: 'i', tone: Some(1) }),
        ("í", ToneChar { base: 'i', tone: Some(2) }),
        ("ǐ", ToneChar { base: 'i', tone: Some(3) }),
        ("ì", ToneChar { base: 'i', tone: Some(4) }),
        ("ō", ToneChar { base: 'o', tone: Some(1) }),
        ("ó", ToneChar { base: 'o', tone: Some(2) }),
        ("ǒ", ToneChar { base: 'o', tone: Some(3) }),
        ("ò", ToneChar { base: 'o', tone: Some(4) }),
        ("ū", ToneChar { base: 'u', tone: Some(1) }),
        ("ú", ToneChar { base: 'u', tone: Some(2) }),
        ("ǔ", ToneChar { base: 'u', tone: Some(3) }),
        ("ù", ToneChar { base: 'u', tone: Some(4) }),
        ("ǖ", ToneChar { base: 'ü', tone: Some(1) }),
        ("ǘ", ToneChar { base: 'ü', tone: Some(2) }),
        ("ǚ", ToneChar { base: 'ü', tone: Some(3) }),
        ("ǜ", ToneChar { base: 'ü', tone: Some(4) }),
    ]
}

fn tone_marks() -> &'static [(&'static str, [&'static str; 5])] {
    &[
        ("a", ["", "ā", "á", "ǎ", "à"]),
        ("e", ["", "ē", "é", "ě", "è"]),
        ("i", ["", "ī", "í", "ǐ", "ì"]),
        ("o", ["", "ō", "ó", "ǒ", "ò"]),
        ("u", ["", "ū", "ú", "ǔ", "ù"]),
        ("v", ["", "ǖ", "ǘ", "ǚ", "ǜ"]),
    ]
}

fn has_marked_vowel(s: &str) -> bool {
    s.chars().any(|c| "āáǎàēéěèīíǐìōóǒòūúǔùǖǘǚǜ".contains(c))
}

fn tone_vowel_index(base: &str) -> isize {
    let s = base.to_lowercase().replace('v', "ü");
    if let Some(i) = s.find('a') {
        return i as isize;
    }
    if let Some(i) = s.find('e') {
        return i as isize;
    }
    if let Some(i) = s.find("ou") {
        return i as isize;
    }
    if let Some(i) = s.find("iu") {
        return i as isize;
    }
    if let Some(i) = s.find("ui") {
        return i as isize;
    }
    let chars: Vec<char> = s.chars().collect();
    for i in (0..chars.len()).rev() {
        let ch = chars[i];
        if "aeiouü".contains(ch) {
            return i as isize;
        }
    }
    -1
}

fn mark_syllable(syl: &str) -> String {
    let raw = syl.trim();
    if raw.is_empty() || has_marked_vowel(raw) {
        return raw.to_string();
    }
    let lower = raw.to_lowercase();
    let re = regex_like_digit_tone(&lower);
    if let Some((base, tone)) = re {
        if tone == 5 {
            return base.replace('v', "ü");
        }
        let idx = tone_vowel_index(&base);
        if idx < 0 {
            return base.replace('v', "ü");
        }
        let chars: Vec<char> = base.chars().collect();
        let ch = chars[idx as usize];
        let key = if ch == 'ü' { 'v' } else { ch };
        if let Some((_, marks)) = tone_marks().iter().find(|(k, _)| *k == key.to_string()) {
            let marked = marks[tone as usize];
            let mut out = String::new();
            for (i, c) in chars.iter().enumerate() {
                if i as isize == idx {
                    out.push_str(marked);
                } else {
                    out.push(*c);
                }
            }
            return out;
        }
        return base.replace('v', "ü");
    }
    raw.to_string()
}

fn regex_like_digit_tone(lower: &str) -> Option<(String, u8)> {
    if lower.len() < 2 {
        return None;
    }
    let last = lower.chars().last()?;
    if !"12345".contains(last) {
        return None;
    }
    let tone: u8 = last.to_digit(10)? as u8;
    let base = lower[..lower.len() - 1].to_string();
    Some((base, tone))
}

pub fn normalize_pinyin_input(pinyin: &str) -> String {
    pinyin
        .trim()
        .split_whitespace()
        .filter(|s| !s.is_empty())
        .map(mark_syllable)
        .collect::<Vec<_>>()
        .join(" ")
}

pub fn normalize_entry(item: &IdiomInput) -> Option<IdiomInput> {
    let word = item.word.trim().to_string();
    let pinyin = normalize_pinyin_input(&item.pinyin);
    if word.is_empty() || pinyin.is_empty() {
        return None;
    }
    Some(IdiomInput {
        word,
        pinyin,
        explanation: item.explanation.clone(),
    })
}

fn strip_tone(syl: &str) -> (String, i32) {
    let mut s = syl.to_lowercase().replace('ü', "v");
    let mut tone = 0i32;

    if let Some(last) = s.chars().last() {
        if "12345".contains(last) {
            tone = last.to_digit(10).unwrap_or(5) as i32;
            s.pop();
        }
    }

    for (marked, tc) in tone_chars() {
        if s.contains(marked) {
            s = s.replace(marked, &tc.base.to_string());
            if let Some(t) = tc.tone {
                tone = t as i32;
            }
            break;
        }
    }

    if tone == 0 {
        tone = 5;
    }

    (s, tone)
}

fn split_initial(py: &str) -> (String, String) {
    let lower = py.to_lowercase();
    let mut initials: Vec<&str> = INITIALS.to_vec();
    initials.sort_by_key(|b| std::cmp::Reverse(b.len()));
    for init in initials {
        if lower.starts_with(init) {
            return (init.to_string(), lower[init.len()..].to_string());
        }
    }
    (String::new(), lower)
}

pub fn parse_syllable(syl: &str) -> CharData {
    let (base, tone) = strip_tone(syl.trim());
    let (sm, ym) = split_initial(&base);
    CharData {
        c: String::new(),
        py: base,
        sm,
        ym,
        tone,
    }
}

pub fn build_chars(word: &str, pinyin: &str) -> Option<Vec<CharData>> {
    let chars: Vec<char> = word.chars().collect();
    let syls: Vec<&str> = pinyin.split_whitespace().filter(|s| !s.is_empty()).collect();
    if chars.len() != syls.len() {
        return None;
    }
    Some(
        chars
            .iter()
            .zip(syls.iter())
            .map(|(c, syl)| {
                let mut parsed = parse_syllable(syl);
                parsed.c = c.to_string();
                parsed
            })
            .collect(),
    )
}

fn all_distinct<T: Eq + std::hash::Hash + Clone>(values: &[T]) -> bool {
    let set: std::collections::HashSet<_> = values.iter().cloned().collect();
    set.len() == values.len()
}

pub fn is_blind_eligible(chars: &[CharData]) -> bool {
    if chars.is_empty() {
        return false;
    }
    all_distinct(&chars.iter().map(|c| c.c.clone()).collect::<Vec<_>>())
        && all_distinct(&chars.iter().map(|c| c.sm.clone()).collect::<Vec<_>>())
        && all_distinct(&chars.iter().map(|c| c.ym.clone()).collect::<Vec<_>>())
        && all_distinct(&chars.iter().map(|c| c.tone).collect::<Vec<_>>())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_dan4() {
        let ch = parse_syllable("dan4");
        assert_eq!(ch.sm, "d");
        assert_eq!(ch.ym, "an");
        assert_eq!(ch.tone, 4);
    }

    #[test]
    fn normalize_marked_and_digit() {
        assert_eq!(
            normalize_pinyin_input("xián zhāng chě jù"),
            "xián zhāng chě jù"
        );
        assert_eq!(
            normalize_pinyin_input("xian2 zhang1 che3 ju4"),
            "xián zhāng chě jù"
        );
    }
}
